codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerDowngrader.java

```diff
@@ -0,0 +1,119 @@
+        if (frame instanceof Http2HeadersFrame) {
+            int id = 0; // not really the id
+            Http2HeadersFrame headersFrame = (Http2HeadersFrame) frame;
+            Http2Headers headers = headersFrame.headers();
+
+            if (headersFrame.isEndStream()) {
+                if (headers.method() == null) {
+                    LastHttpContent last = new DefaultLastHttpContent(Unpooled.EMPTY_BUFFER, true);
+                    HttpConversionUtil.addHttp2ToHttpHeaders(id, headers, last.trailingHeaders(),
+                                                             HttpVersion.HTTP_1_1, true, true);
+                    out.add(last);
+                } else {
+                    FullHttpRequest full = HttpConversionUtil.toHttpRequest(0, headers, ctx.alloc(), true);
+                    out.add(full);
+                }
+            } else {
+                out.add(HttpConversionUtil.toHttpChunkedRequest(id, headersFrame.headers(), true));
+            }
+
+        } else if (frame instanceof Http2DataFrame) {
+            Http2DataFrame dataFrame = (Http2DataFrame) frame;
+            if (dataFrame.isEndStream()) {
+                out.add(new DefaultLastHttpContent(dataFrame.content(), true));
+            } else {
+                out.add(new DefaultHttpContent(dataFrame.content()));
+            }
+        }
+        ReferenceCountUtil.retain(frame);
+    }
+
+    private static void encodeLastContent(LastHttpContent last, List<Object> out) {
+        boolean needFiller = !(last instanceof FullHttpResponse) && last.trailingHeaders().isEmpty();
+        if (last.content().readableBytes() > 0 || needFiller) {
+            out.add(new DefaultHttp2DataFrame(last.content(), last.trailingHeaders().isEmpty()));
+        }
+        if (!last.trailingHeaders().isEmpty()) {
+            Http2Headers headers = HttpConversionUtil.toHttp2Headers(last.trailingHeaders(), true);
+            out.add(new DefaultHttp2HeadersFrame(headers, true));
+        }
+    }
+
+    @Override
+    protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> out) throws Exception {
+        if (obj instanceof FullHttpResponse) {
+            FullHttpResponse full = (FullHttpResponse) obj;
+            Http2Headers headers = HttpConversionUtil.toHttp2Headers(full, true);
+            boolean noMoreFrames = full.content().readableBytes() == 0 && full.trailingHeaders().isEmpty();
+            out.add(new DefaultHttp2HeadersFrame(headers, noMoreFrames));
+            if (!noMoreFrames) {
+                encodeLastContent(full, out);
+            }
+
+        } else if (obj instanceof HttpResponse) {
+            HttpResponse req = (HttpResponse) obj;
+            Http2Headers headers = HttpConversionUtil.toHttp2Headers(req, true);
+            out.add(new DefaultHttp2HeadersFrame(headers, false));
+
+        } else if (obj instanceof LastHttpContent) {
```

Comment:
LastHttpContent is not mutually exclusive with HttpResponse (as is seen with FullHttpResponse). Instead of if-elses, I'd suggest having separate ifs for each part of the message, and don't treat FullHttpResponse specially.

```
if (obj instanceof HttpMessage) {
  out.add(headers);
}
if (obj instanceof HttpContent) {
  if (hasContent) {
    out.add(content /* no eos */);
  }
}
if (obj instanceof LastHttpContent) {
  if (lastContent.trailingHeaders().isEmpty()) {
    out.add(emptyDataFrameWithEOS);
  } else {
    out.add(trailers);
  }
}
```

It would be possible to optimize it to avoid `emptyDataFrameWithEOS`, but I question how much that matters.